org.lflang/src/org/lflang/graph/InstantiationGraph.java

```diff
@@ -0,0 +1,178 @@
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.lflang.graph;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Iterables;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.eclipse.emf.common.util.EList;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.xtext.xbase.lib.CollectionLiterals;
+import org.eclipse.xtext.xbase.lib.Functions.Function1;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+import org.eclipse.xtext.xbase.lib.IteratorExtensions;
+import org.lflang.ASTUtils;
+import org.lflang.lf.Instantiation;
+import org.lflang.lf.Model;
+import org.lflang.lf.Reactor;
+import org.lflang.lf.ReactorDecl;
+
+/**
+ * A graph with vertices that are Reactors (not ReactorInstances) and edges that denote
+ * dependencies between them. A "dependency" from reactor class A to
+ * reactor class B (A depends on B) means that A instantiates within
+ * it at least one instance of B. Note that there a potentially
+ * confusing and subtle distinction here between an "instantiation"
+ * and an "instance". They are not the same thing at all. An
+ * "instantiation" is an AST node representing a statement like
+ * `a = new A();`. This can result in many instances of reactor
+ * class A (if the containing reactor class is instantiated multiple times).
+ *
+ * In addition to the graph, this class keeps track of the instantiations
+ * that induce the dependencies. These can be retrieved using the method
+ * `getInstantiations(Reactor)`.
+ *
+ * @author{Marten Lohstroh <marten@berkeley.edu>}
+ */
+public class InstantiationGraph extends PrecedenceGraph<Reactor> {
+    /**
+     * A mapping from reactors to the sites of their instantiation.
+     */
+    protected final HashMultimap<Reactor, Instantiation> reactorToInstantiation =
+        HashMultimap.<Reactor, Instantiation>create();
+
+    /**
+     * A mapping from reactor classes to their declarations.
+     */
+    protected final HashMultimap<Reactor, ReactorDecl> reactorToDecl =
+        HashMultimap.<Reactor, ReactorDecl>create();
+
+    protected final HashSet<Object> flaggedReactors = CollectionLiterals.<Object>newHashSet();
```

Comment:
This looks to be unused. Should we remove it?

Reply:
Great catch!